### DISCUSSION TITLE
Adding extralibs

### DISCUSSION THREAD

**Discussion by vitor-txt** [CodinGame/monaco-vscode-api#498] [Q&A] [answered]
Hi! In the past, when I used the Monaco Editor, I was able to add `.d.ts` files to JavaScript with the command `monaco.languages.typescript.javascriptDefaults.addExtraLib(libSource, libUri)`, similar to what is done here:
https://microsoft.github.io/monaco-editor/playground.html?source=v0.51.0#example-extending-language-services-configure-javascript-defaults

However, for some reason, when I try to perform this action, I am getting `undefined` for `languages.typescript`. I have also tried:
- Importing directly to `monaco-editor`: 
```js
import * as monaco from 'monaco-editor';
//same with import * as editor from 'monaco-editor/esm/vs/editor/editor.api';
...
monaco.languages.typescript.javascriptDefaults.addExtraLib(lib)

```
- Using `getApi`:
```js
  void getApi().then(async (api: vscodeAPI) => {
      ...
      const extraLib = api.languages.typescript.javascriptDefaults.addExtraLib(lib);
```

Am I missing something, or is there another way to add libraries?

**Comment by CGNonofr** [comment, marked as answer]
Hello,

Those apis are added by the monaco standalone typescript worker.

It that project, we tend to encourage using VSCode stuff instead, and you should be able to use the VSCode typescript default extension for the typescript intellisense (cf. [the demo](https://monaco-vscode-api.netlify.app/?mode=full-workbench)). You would be able to add types but adding your `.d.ts` file in the project directly in the virtual filesystem, or referencing published types (there is a type acquisition feature).

However, there is something that can be blocking for many users: the VSCode typescript extension requires the `crossOriginIsolated` flag to be enable to fully operate (aka `project-wide` intellisense).

For that reason, it's still possible to use the monaco typescript worker, you just need to install and import `@codingame/monaco-vscode-standalone-typescript-language-features`

**Reply by Nuthin2do** [reply]
> For that reason, it's still possible to use the monaco typescript worker, you just need to install and import @codingame/monaco-vscode-standalone-typescript-language-features

I have a question regarding this. I installed and imported the lib. On creation of the editor and its model i do something like this:

```
import "@codingame/monaco-vscode-standalone-typescript-language-features";

...

public setTypescriptValidationOptions() {
    const typescriptDefaults = languages.typescript.typescriptDefaults;
    typescriptDefaults.setDiagnosticsOptions({
      noSemanticValidation: false,
      noSyntaxValidation: false,
    });
    const tsOp = typescriptDefaults.getCompilerOptions();
    tsOp.allowJs = true;
    tsOp.checkJs = true;
    tsOp.target = languages.typescript.ScriptTarget.ES2020;
    tsOp.lib = ["ES6"];
    tsOp.isolatedModules = true;
    typescriptDefaults.setCompilerOptions(tsOp);
  }

  public setDtsFiles(dtsFiles: DtsFile | DtsFile[]) {
    if (Array.isArray(dtsFiles)) {
      const extraLibs = dtsFiles.map((x) => ({
        content: x.content,
        filePath: x.name,
      }));
      languages.typescript.javascriptDefaults.setExtraLibs(extraLibs);
    } else {
      const tempDtsFiles = dtsFiles;
      const extraLibs = [
        { content: tempDtsFiles.content, filePath: tempDtsFiles.name },
      ];
      languages.typescript.javascriptDefaults.setExtraLibs(extraLibs);
    }
  }
```

But as soon as i start the editor i get these errors on startup and on each keystroke:

<img width="347" alt="image" src="https://github.com/user-attachments/assets/8a3f4ce6-db39-40c0-8258-1451f1902e10">

Can you tell me what i am missing? If you need some more information, please tell me.

**Reply by CGNonofr** [reply]
The language features uses a worker and need you to register it by returning a worker on `@codingame/monaco-vscode-standalone-typescript-language-features/worker` in the `getWorker` function for label `typescript` and `javascript`

**Reply by Nuthin2do** [reply]
Hi, many thanks for your answer.

I assume this is not the way to go, is it? (The import and usage is as it was)

```
type WorkerLoader = () => Worker;

    const workerLoader: Partial<Record<string, WorkerLoader>> = {
      TextEditorWorker: () =>
        new Worker(
          new URL(
            "monaco-editor/esm/vs/editor/editor.worker.js",
            import.meta.url
          ),
          { type: "module" }
        ),
      TextMateWorker: () =>
        new Worker(
          new URL(
            "@codingame/monaco-vscode-textmate-service-override/worker",
            import.meta.url
          ),
          { type: "module" }
        ),
    };

    window.MonacoEnvironment = {
      getWorker: function (moduleId, label) {
        const workerFactory = workerLoader[label];
        if (label === "typescript" || label === "javascript") {
          return new Worker(
            new URL(
              "@codingame/monaco-vscode-standalone-typescript-language-features/worker",
              import.meta.url
            ),
            { type: "module" }
          );
        } else if (workerFactory != null) {
          return workerFactory();
        } else {
          throw new Error(`Unimplemented worker ${label} (${moduleId})`);
        }
      },
    };
```

Because i still get errors although this time the errors are different:

<img width="369" alt="image" src="https://github.com/user-attachments/assets/8f7e8898-2bb3-4545-80da-d758ce403961">

Can you tell me what exactly i am doing wrong?

**Reply by CGNonofr** [reply]
It looks good to me

what bundler are you using?

**Reply by Nuthin2do** [reply]
I am using vite and this is the config (mostly a copy of the one from the demo)

```
import { defineConfig } from "vite";
import importMetaUrlPlugin from "@codingame/esbuild-import-meta-url-plugin";
import pkg from "./package.json" with { type: "json" };
import path from "path";

// Extract local dependencies (if any are relevant)
const localDependencies = Object.entries(pkg.dependencies)
  .filter(([, version]) => version.startsWith("file:../"))
  .map(([name]) => name);

export default defineConfig({
  build: {
    target: "esnext",
    outDir: path.resolve(__dirname, "..", "wwwroot", "my-editor"),
    rollupOptions: {
      input: path.resolve(__dirname, "src/index.ts"),
      output: {
        chunkFileNames: "js/[name].js",
        entryFileNames: "js/[name].js",
        assetFileNames: "[ext]/[name].[ext]",
      },
    },
  },
  worker: {
    format: "es",
  },
  optimizeDeps: {
    include: [
      ...localDependencies,
      "vscode/extensions",
      "vscode/services",
      "vscode/monaco",
      "vscode/localExtensionHost",
      "vscode-textmate",
      "vscode-oniguruma",
      "@vscode/vscode-languagedetection",
      "marked",
    ],
    exclude: [],
    esbuildOptions: {
      tsconfig: "./tsconfig.json",
      plugins: [importMetaUrlPlugin],
    },
  },
  resolve: {
    dedupe: ["vscode", ...localDependencies],
  },
  server: {
    port: 8080,
    host: "localhost",
    fs: {
      allow: ["../"],
    },
  },
  define: {
    rootDirectory: JSON.stringify(__dirname),
  },
});

```

**Reply by CGNonofr** [reply]
If you want me to have a look the, please provide a minimal reproduction repo

**Reply by Nuthin2do** [reply]
[NewEditor.zip](https://github.com/user-attachments/files/17553752/NewEditor.zip)

Hi, this is a basic example aka my first attempt in trying out your repo, but nonetheless i have the same issue in this repo. Please let me know if you need some more information. Thank you very much for your help.

**Reply by CGNonofr** [reply]
I've just had a look, and I've noticed the monaco-editor exposed worker api was broken in the last VSCode version

No monaco-editor version was released for a long time now, I guess they'll fix it

The fix doesn't seem to complicated though => https://github.com/CodinGame/monaco-vscode-api/pull/526

**Reply by CGNonofr** [reply]
can you have a look at the v10.1.2?

btw, importing VSCode theme extension without using the theme service override doesn't make sense

**Reply by CGNonofr** [reply]
(Did you remove your message or is it a github bug?)

**Reply by Nuthin2do** [reply]
[NewEditor.zip](https://github.com/user-attachments/files/17560222/NewEditor.zip)

Unfortunately there is still an error. If i run `npm run start` then the browser shows this error:

<img width="903" alt="image" src="https://github.com/user-attachments/assets/284787b0-72de-4462-9eaf-195c3b3a1326">

But when i run `npm run build` then it doesn't finish and this error pops up:

```
error during build:
[vite:worker-import-meta-url] node_modules/@codingame/monaco-vscode-standalone-typescript-language-features/worker.js (2:9): "initialize" is not exported by "node_modules/monaco-editor/esm/vs/editor/editor.worker.js", imported by "node_modules/@codingame/monaco-vscode-standalone-typescript-language-features/worker.js".
file: C:/temp/NewEditor/src/main.ts:2:9

1: import { initialize } from 'monaco-editor/esm/vs/editor/editor.worker.js';
2: export { initialize } from 'monaco-editor/esm/vs/editor/editor.worker.js';
            ^
3:
4: /*!-----------------------------------------------------------------------------

at getRollupError (file:///C:/temp/NewEditor/node_modules/rollup/dist/es/shared/parseAst.js:395:41)
    at error (file:///C:/temp/NewEditor/node_modules/rollup/dist/es/shared/parseAst.js:391:42)
    at Module.error (file:///C:/temp/NewEditor/node_modules/rollup/dist/es/shared/node-entry.js:15551:16)
    at Module.traceVariable (file:///C:/temp/NewEditor/node_modules/rollup/dist/es/shared/node-entry.js:16000:29)
    at ModuleScope.findVariable (file:///C:/temp/NewEditor/node_modules/rollup/dist/es/shared/node-entry.js:13786:39)
    at ReturnValueScope.findVariable (file:///C:/temp/NewEditor/node_modules/rollup/dist/es/shared/node-entry.js:5252:38)
    at FunctionBodyScope.findVariable (file:///C:/temp/NewEditor/node_modules/rollup/dist/es/shared/node-entry.js:5252:38)
    at Identifier.bind (file:///C:/temp/NewEditor/node_modules/rollup/dist/es/shared/node-entry.js:5035:40)
    at CallExpression.bind (file:///C:/temp/NewEditor/node_modules/rollup/dist/es/shared/node-entry.js:2855:23)
    at CallExpression.bind (file:///C:/temp/NewEditor/node_modules/rollup/dist/es/shared/node-entry.js:11251:15)
```

See the repo with the updated packages.

> btw, importing VSCode theme extension without using the theme service override doesn't make sense

If i don't use this import:

`import "@codingame/monaco-vscode-theme-defaults-default-extension";`

in combination with this:

```
await initialize(
    {
      ...getModelServiceOverride(),
      ...getThemeServiceOverride(),
      ...getLanguagesServiceOverride(),
      ...getTextMateServiceOverride(),
      ...getConfigurationServiceOverride(),
      ...getDebugServiceOverride(),
    },
    undefined,
```

then there is no highlighting and changing the theme in the config.json doesn't change anything. Therefore i'm a bit confused what you mean.... I’d appreciate any tips on this. Thank you

**Reply by CGNonofr** [reply]
Sorry, I made a last-minute refactor that broke it again in a different way... fixed in https://github.com/CodinGame/monaco-vscode-api/pull/527

Regarding the theme service, I was convinced you were'nt using it, I probably misread :shrug:

**Reply by Nuthin2do** [reply]
Now everything works like a charm. Thank you very much 👍
